### ISSUE TITLE
"wxLoadUserResource Fails to Load Integer Resource IDs on Windows Due to Incorrect Use of FindResource

### ISSUE THREAD

**Issue by leonid-galperin** [wxWidgets/wxWidgets#24947]
#### Bug description:

In **wxWidgets 3.2.6** on Windows, the function `wxLoadUserResource` cannot load resources using an integer resource ID. Currently, the implementation of `wxLoadUserResource` in `utils.cpp` calls `FindResource` directly with `resourceName.t_str()`:

```cpp
HRSRC hResource = ::FindResource(instance, resourceName.t_str(), resourceType);
```

This causes an issue because `resourceName.t_str()` is treated as a string, making it impossible to use `wxLoadUserResource` with an integer resource ID as expected in cases where `MAKEINTRESOURCE` is necessary for the `FindResource` call.

#### Expected vs observed behavior:

**Expected behavior:**

- `wxLoadUserResource` should be able to accept integer resource IDs and pass them correctly to `FindResource` using `MAKEINTRESOURCE` when needed.

**Observed behavior:**

- Currently, `wxLoadUserResource` fails when an integer ID is passed, as `FindResource` does not interpret it as a resource ID.

#### Stack trace:

N/A (no crash, but resource loading fails for integer IDs).

#### Patch or snippet allowing to reproduce the problem:

Example call that fails due to the current `wxLoadUserResource` implementation:

```cpp
const void* outData = nullptr;
size_t outLen = 0;
HINSTANCE hInstance = GetModuleHandle(nullptr);

if (!wxLoadUserResource(&outData, &outLen, "165", wxT("RCDATA"), hInstance)) {
    wxLogError("Failed to load resource.");
}
```

**Proposed fix:** Modify `wxLoadUserResource` to accept an integer `resourceId` instead of a `wxString` for `resourceName`. Implement `MAKEINTRESOURCE` for integer resource IDs.

Suggested updated function signature:

```cpp
bool wxLoadUserResource(const void** outData, size_t* outLen, int resourceId, const wxChar* resourceType, WXHINSTANCE instance);
```

Updated call to `FindResource`:

```cpp
HRSRC hResource = ::FindResource(instance, MAKEINTRESOURCE(resourceId), resourceType);
```

Alternatively, the function could support both `wxString` and integer IDs by adding an overload or using an optional parameter type.

### Platform and version information

- **wxWidgets version**: 3.2.6
- **wxWidgets port**: wxMSW
- **OS**: Windows 10
  
  - GTK version: 
  - Which GDK backend is used: 
  - Desktop environment: 
  - Current theme:

**Comment by PBfordev** [comment]
This is a known long-time deficiency of `wxLoadUserResource()`.

Fortunately, it is trivial to work around, calling it using the resource name pattern starting with `#`.

For example, if your .rc file has something like this
```
#include "resource.h"

IDB_STAR RCDATA "star.png"
```
and file `resource.h` has
```
#define IDB_STAR 101
```

the correct `resourceName` argument for calling `wxLoadUserResource()` can be obtained with `wxString::Format(wxString::Format("#%d", IDB_STAR);`

**Comment by PBfordev** [comment]
> Well.. First of all thanks! Didn't know that. 

Perhaps the workaround could be mentioned in `wxLoadUserResource()` docs.

> Anyway it doesn't help for loading buttons/toolbar tools bitmaps from resource as they ask for resource name and do not convert them to id with '#'.

Who is "they" here? AFAIK wxWidgets controls do not use resource ids, they use `wxBitmap{Bundle}` etc. where the workaround works when loading from a Windows resource. If you mean macros like `wxBITMAP()` then I guess it may be possible to use some preprocessor magic to create a helper macro converting from integer id but I do not know how. 

But the macros are not related to the Issue, i.e., an alleged bug in `wxLoadUserResource()`.
 
> And second - since it "known long-time", probably it's time to fix it?

TBH, I don't see a bug and hence nothing to fix, only a possible enhancement. In my experience, in wxWorld (contrary to e.g. MFC), integers are rarely used as resource identifiers and that is why it works as it does. Of course, you are free to submit a nice backward compatible PR enhancing `wxLoadUserResource()` and see what the devs will say.

**Comment by vadz** [comment]
Just to confirm the above, I wouldn't mind applying a PR adding support for integer IDs or even a PR improving the documentation to explain the current workaround, of course, but I have no plans to change this myself as I never use integer IDs personally.

**Comment by PBfordev** [comment]
My two cents:

AFAICT, the bug described in the OP is not a bug, a string with value `"123"` is not the same as int with value `123`. `MAKEINTRESOURCE()` does not actually use a string, it uses a hack where an address pointing to within the first 64 kB of heap is treated as an integer and not pointer to `TCHAR` string.

We cannot just change the type of `resourceName` argument of `wxLoadUserResource()` to `char*`/`wxChar*`, which could in theory be called with both `MAKEINTRESOURCE()` and a string literal/`wxString`. This would break not only due to Unicode/non-Unicode literals but also when building with `wxUSE_UNSAFE_WXSTRING_CONV` set to `1`.

We can add `wxLoadUserResource()` overload taking an `int` instead of `wxString&` as the resource id; however, this will not make macros like `wxBITMAP` work with integer identifiers. I think the macros is the real issue here, when calling `wxLoadUserResource()` directly, it is trivial to use the workaround (if one is aware of it). Therefore, there may not be a good enough reason to do anything about this.

**Comment by vadz** [comment]
Yes, I understood the "bug" as actually being a request for enhancement to provide an overload of this function taking `int`, I didn't think about `wxBITMAP`. IMO it's not worth changing the latter, but the former could still be implemented if anybody really cares about it.
